server/src/main/java/org/opensearch/action/search/SearchRequestContext.java

```diff
@@ -39,6 +39,10 @@ public class SearchRequestContext {
     private TotalHits totalHits;
     private final EnumMap<ShardStatsFieldNames, Integer> shardStats;
     private Set<Index> successfulSearchShardIndices;
+    private Map<SearchPhase, Boolean> phaseStartedMap;
+    private Map<SearchPhase, Boolean> phaseEndedMap;
```

Comment:
This will bloat quickly and can have serious memory implications. I think single structure should be enough. Keep a single set and verify  for both start and end events. Ideally we should remove the phase value on end event.

Two or more phases can start concurrently hence the hashmap will throw concurrent modification exception